site/student-lectures/cultural-algorithms/ca.py

```diff
@@ -0,0 +1,141 @@
+import math
+import random
+import numpy as np
+
+POPULATION_SIZE = 30
+GENERATIONS = 100
+ACCEPT_NUMBER = 2
+
+def fitness_function(individual):
+    return individual[0] ** 2 + individual[1] ** 2
+
+def initialize_population():
+    population = []
+    for _ in range(POPULATION_SIZE):
+        individual = [random.uniform(-10, 10) for _ in range(2)]
+        population.append(individual)
+    return population
+
+
+def initialize_belief_space():
+    # initialize situational
+    belief_space = {"situational": None, "normative": []}
+    # initialize normative
+    for _ in range(2):
+        # [min, max, Lower, Upper]
+        belief_space["normative"].append([-10, 10, 100, 100])
+    return belief_space
+
+
+
+def accept(population, population_fitness):
+    combined_data = list(zip(population, population_fitness))
+    sorted_combined_data = sorted(combined_data, key=lambda x: x[1])
+    sorted_population, sorted_population_fitness = zip(*sorted_combined_data)
+    # ACCEPT_NUMBER is a constant
+    return sorted_population[:ACCEPT_NUMBER]
+
+
+def update(belief_space, n_accepted):
+    for x_l in n_accepted:
+        # update situational
+        if belief_space["situational"] is None:
+            belief_space["situational"] = x_l
+        elif fitness_function(x_l) < fitness_function(belief_space["situational"]):
+            belief_space["situational"] = x_l
+
+        # update_normative
+        for j in range(2):
+            bound = belief_space["normative"][j]
+            # update min and L
+            if x_l[j] <= bound[0] or fitness_function(x_l) < bound[2]:
+                bound[0] = x_l[j]
+                bound[2] = fitness_function(x_l)
+            # update max and U
+            if x_l[j] >= bound[1] or fitness_function(x_l) < bound[3]:
+                bound[1] = x_l[j]
+                bound[3] = fitness_function(x_l)
+
+    print(belief_space)
```

Comment:
don't print. return. If you want to print it out, print it out after it has been returned to where it was called. 

Also, needs tests.